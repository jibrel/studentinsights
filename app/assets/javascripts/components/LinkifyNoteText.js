import PropTypes from 'prop-types';
import React from 'react';
import _ from 'lodash';


// Adapted from NoteText.js
const noteText = {
  marginTop: 5,
  padding: 0,
  fontFamily: "'Open Sans', sans-serif",
  fontSize: 14,
  whiteSpace: 'pre-wrap',
  wordBreak: 'break-word',
  overflowX: 'hidden'
};


export default function LinkifyNoteText(props) {
  const {text, safeDomains, style} = props;
  const urls = _findURLsWithinText(safeDomains, text);
  const pieces = _join(text, urls);
  return (
    <div className="LinkifyNoteText" style={style || {}}>
      {pieces.map((piece, index) => {
        if (piece.text) {
          return <span key={index} style={noteText}>{piece.text}</span>;
        }

        if (piece.url) {
          return (
            <a
              key={index}
              href={piece.url}
              style={noteText}
              target="_blank"
              rel="noopener noreferrer"
            >
              {piece.url}
            </a>
          );
        }

        return null;
      })}
    </div>
  );
}

LinkifyNoteText.propTypes = {
  text: PropTypes.string.isRequired,
  style: PropTypes.object,
  safeDomains: PropTypes.arrayOf(PropTypes.string).isRequired
};
LinkifyNoteText.defaultProps = {
  safeDomains: [
    'docs.google.com',
    'drive.google.com'
  ]
};


// Exported only for tests
export function _join(text, urls) {
  if (text.length === 0) return [];

  // find first url
  const matches = urls.map(url => {
    return {url, index: text.indexOf(url)};
  });
  const match = _.min(matches.filter(match => match.index !== -1), 'index');
  if (!match) return [{text}];

  // split text
  const textBefore = text.slice(0, match.index);
  const textAfter = text.slice(match.index + match.url.length);

  // recur
  return _.compact([
    (textBefore.length > 0) ? {text: textBefore } : null,
    {url: match.url},
  ]).concat(_join(textAfter, urls));
}

// Exportd only for testing
export function _findURLsWithinText(safeDomains, text) {
  return _.flatMap(safeDomains, safeDomain => {
    const regex = new RegExp('https://' + escapeToUseAsRegexLiteral(safeDomain) + '[^\\s]*');
    const match = text.match(regex);
    return (match) ? match : [];
  });
}

function escapeToUseAsRegexLiteral(x) {
  return x.replace('?', '\\?');
}
