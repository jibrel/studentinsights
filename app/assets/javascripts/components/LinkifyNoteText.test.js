import React from 'react';
import renderer from 'react-test-renderer';
import LinkifyNoteText, {_findURLsWithinText, _join} from './LinkifyNoteText';


function testProps(props = {}) {
  return {
    text: 'hello',
    safeDomains: ['example.com', 'sub.test.com'],
    ...props
  };
}

function testCases() {
  return [
    testProps({text: "hello https://example.com/page?query=yes goodbye"}),
    testProps({text: "hello https://example.com/page?query=yes+please#also-this goodbye"}),
    testProps({text: "hello https://example.com/ goodbye"}),
    testProps({text: "hello https://example.com goodbye"}),
    testProps({text: "hello https://sub.example.com goodbye"}),
    testProps({text: "hello http://example.com goodbye"}),
    testProps({text: "hello example.com goodbye"}),
    testProps({text: "hello https://sub.test.com/ goodbye"}),
    testProps({text: "hello https://test.com/ goodbye"})
  ];
}

it('#_findURLsWithinText works', () => {
  const safeDomains = [
    'example.com',
    'sub.test.com'
  ];
  expect(_findURLsWithinText(safeDomains, 'hello https://example.com/page goodbye')).toEqual(['https://example.com/page']);
  expect(_findURLsWithinText(safeDomains, 'hello https://example.com/page?query=yes goodbye')).toEqual(['https://example.com/page?query=yes']);
  expect(_findURLsWithinText(safeDomains, 'hello https://example.com/page?query=yes+please#also-this goodbye')).toEqual(['https://example.com/page?query=yes+please#also-this']);
  expect(_findURLsWithinText(safeDomains, 'hello https://example.com/ goodbye')).toEqual(['https://example.com/']);
  expect(_findURLsWithinText(safeDomains, 'hello https://example.com goodbye')).toEqual(['https://example.com']);
  expect(_findURLsWithinText(safeDomains, 'hello https://sub.example.com goodbye')).toEqual([]);
  expect(_findURLsWithinText(safeDomains, 'hello http://example.com goodbye')).toEqual([]);
  expect(_findURLsWithinText(safeDomains, 'hello example.com goodbye')).toEqual([]);
  expect(_findURLsWithinText(safeDomains, 'hello https://sub.test.com/ goodbye')).toEqual(['https://sub.test.com/']);
  expect(_findURLsWithinText(safeDomains, 'hello https://test.com/ goodbye')).toEqual([]);
});

it('joins', () => {
  function testCase(text) {
    const safeDomains = ['example.com'];
    const urls = _findURLsWithinText(safeDomains, text);
    const pieces = _join(text, urls);
    return pieces;
  }
  

  expect(testCase('hello https://example.com/ goodbye')).toEqual([
    { text: 'hello ' },
    { url: 'https://example.com/' },
    { text: ' goodbye' }
  ]);

  expect(testCase('https://example.com/ goodbye')).toEqual([
    { url: 'https://example.com/' },
    { text: ' goodbye' }
  ]);

  expect(testCase('hello https://example.com/page')).toEqual([
    { text: 'hello ' },
    { url: 'https://example.com/page' }
  ]);
});

it('snapshots across test cases', () => {
  const tree = renderer
    .create(<div>{testCases().map((props, index) => (
      <LinkifyNoteText key={index} {...props} />
    ))}</div>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});