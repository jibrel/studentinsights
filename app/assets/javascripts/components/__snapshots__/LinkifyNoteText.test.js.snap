// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots across test cases 1`] = `
<div>
  <div
    className="LinkifyNoteText"
    style={Object {}}
  >
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
      hello 
    </span>
    <a
      href="https://example.com/page?query=yes"
      rel="noopener noreferrer"
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
      target="_blank"
    >
      https://example.com/page?query=yes
    </a>
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
       goodbye
    </span>
  </div>
  <div
    className="LinkifyNoteText"
    style={Object {}}
  >
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
      hello 
    </span>
    <a
      href="https://example.com/page?query=yes+please#also-this"
      rel="noopener noreferrer"
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
      target="_blank"
    >
      https://example.com/page?query=yes+please#also-this
    </a>
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
       goodbye
    </span>
  </div>
  <div
    className="LinkifyNoteText"
    style={Object {}}
  >
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
      hello 
    </span>
    <a
      href="https://example.com/"
      rel="noopener noreferrer"
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
      target="_blank"
    >
      https://example.com/
    </a>
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
       goodbye
    </span>
  </div>
  <div
    className="LinkifyNoteText"
    style={Object {}}
  >
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
      hello 
    </span>
    <a
      href="https://example.com"
      rel="noopener noreferrer"
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
      target="_blank"
    >
      https://example.com
    </a>
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
       goodbye
    </span>
  </div>
  <div
    className="LinkifyNoteText"
    style={Object {}}
  >
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
      hello https://sub.example.com goodbye
    </span>
  </div>
  <div
    className="LinkifyNoteText"
    style={Object {}}
  >
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
      hello http://example.com goodbye
    </span>
  </div>
  <div
    className="LinkifyNoteText"
    style={Object {}}
  >
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
      hello example.com goodbye
    </span>
  </div>
  <div
    className="LinkifyNoteText"
    style={Object {}}
  >
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
      hello 
    </span>
    <a
      href="https://sub.test.com/"
      rel="noopener noreferrer"
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
      target="_blank"
    >
      https://sub.test.com/
    </a>
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
       goodbye
    </span>
  </div>
  <div
    className="LinkifyNoteText"
    style={Object {}}
  >
    <span
      style={
        Object {
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
      hello https://test.com/ goodbye
    </span>
  </div>
</div>
`;
